LoRa: Add support a shell with support for CW testing

andysan opened the pull request:
The shell provides a handful of commands that are useful for testing:

- lora conf ...
- lora send ...
- lora recv ...
- lora test_cw ...

Tested with an SX1262 module transmitting packets to a b_l072z_lrwan1 board with a builtin SX1276, both running the Zephyr with the LoRa shell. 

~**NOTE:** There is a dependency between #24649 and this pull request. The current version of this pull request assumes that #24649 hasn't been merged.~

**UPDATE**: Dropped support for timeouts in recv command to break the dependency between this  PR  and #24649.

**UPDATE**: Reintroduced support for timeouts in recv since #24649 has been merged. Much improved argument checking.

**UPDATE**: Rebased and switched to the new shell_hexdump to print received packets. Sorry about the noise.

**UPDATE**: Updated to work with new DT_HAS_NODE_STATUS_OKAY macro.

andysan committed:
drivers: lora: Add a continuous wave test API

Add an API to transmit a continuous wave at a fixed frequency. This
functionality is useful to test the radio in a lab setup.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

andysan committed:
drivers: lora: Add a shell for LoRa testing

Provide basic commands that are useful when testing a LoRa
radio. Currently, the shell supports:

> lora conf ...
> lora send ...
> lora recv ...
> lora test_cw ...

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

Mani-Sadhasivam commented on include/drivers/lora.h:
```diff
@@ -139,4 +149,25 @@ static inline int lora_recv(struct device *dev, u8_t *data, u8_t size,
 }
 
+/**
+ * @brief Transmit a continuous wave at a given frequency
```
Need to mention `unmodulated` here. Also, it'd be good to mention the use of this API.

Mani-Sadhasivam commented:
@andysan The shell driver needs to be moved to `samples/drivers/lora`.

Mani-Sadhasivam commented:
Also cleanup the checkpatch errors

andysan commented:
> @andysan The shell driver needs to be moved to `samples/drivers/lora`.

I’m not sure I understand the rationale behind this. Other subsystems, such as GPIO, have shell drivers that live with the drivers themselves. Being able to send a CW or raw packets is very useful when bringing up a new board. Any particular reason why LoRa would be different?

I’m planning to fix the other issues later today or tomorrow. I’m currently trying to find time to debug test and debug the SX126x driver.

Mani-Sadhasivam commented:
> I’m not sure I understand the rationale behind this. Other subsystems, such as GPIO, have shell drivers that live with the drivers themselves. Being able to send a CW or raw packets is very useful when bringing up a new board. Any particular reason why LoRa would be different?

Ah. I presumed that shell belongs to samples (was it initially(?)) and didn't look into other drivers. Now I just checked and all of them live under `drivers/`, so we are good. Sorry for the confusion.

andysan commented:
Question about the shell. How do you prefer "lora config" to work? I'm considering rewriting it so you provide property-value pairs instead of a long list of properties.

Mani-Sadhasivam commented:
> Question about the shell. How do you prefer "lora config" to work? I'm considering rewriting it so you provide property-value pairs instead of a long list of properties.

I think the values alone would be sufficient. Since the help text specifies the property order, the user need not specify the property name again.

andysan commented:
> Also cleanup the checkpatch errors

This should be fixed now. I also improved the error handling in the shell to avoid surprises.

Mani-Sadhasivam commented on drivers/lora/lora_shell.c:
```diff
@@ -0,0 +1,257 @@
+
+SHELL_STATIC_SUBCMD_SET_CREATE(sub_lora,
+	SHELL_CMD_ARG(config, NULL,
+		      "Configure the LoRa radio\n"
```
Please revert to the initial implementation of specifying the values alone.

andysan commented:
> > Question about the shell. How do you prefer "lora config" to work? I'm considering rewriting it so you provide property-value pairs instead of a long list of properties.
> 
> I think the values alone would be sufficient. Since the help text specifies the property order, the user need not specify the property name again.

Having played around with both implementations, I personally prefer the new one since I found myself wanting to change specific parameters without touching the other ones. This is also consistent with how the 'ip' command works on Linux. The 802.15.4 shell in Zephyr seems to have separate get/set commands for every property, but that seems tedious to implement.

As I see it, there are three big benefits: Adding new options doesn't "break" the config command, it's possible to update individual settings, it's less error prone if you want to change a parameter late in the list.

I'm happy to change back if you feel strongly about it though.

Mani-Sadhasivam commented on drivers/lora/lora_shell.c:
```diff
@@ -0,0 +1,257 @@
+		if (i + 1 >= argc) {
+			shell_error(shell, "'%s' expects an argument",
+				    argv[i]);
+			return -1;
```
Return POSIX error code everywhere.

Mani-Sadhasivam commented:
> As I see it, there are three big benefits: Adding new options doesn't "break" the config command, it's possible to update individual settings, it's less error prone if you want to change a parameter late in the list.

Okay. I didn't notice that we can specify individual settings. So we can settle with this, thanks.

andysan commented on drivers/lora/lora_shell.c:
```diff
@@ -0,0 +1,257 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <inttypes.h>
```
Yes, I'm using the PRIuXX / PRIxXX macros.

andysan commented:
I should have addressed all of the issues now. However, I realised that I should have asked if how to select the right radio. I'm currently using `DEFAULT_RADIO_NODE DT_NODELABEL(lora0)` to find a node with the lora0 label. Another option might be to use an alias. Is that a better option?

andysan committed:
boards: arm: 96b_wistrio: Add an alias for the LoRa radio

The LoRa shell and samples need a way to find the default LoRa
radio. Add the DT alias 'lora0' for the default radio.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

andysan committed:
samples: lora: Use the new DT alias to discover radio

The LoRa shell requires the alias 'lora0' for the default radio. Use
the same alias for the samples to make them portable across different
radios and devices.

Signed-off-by: Andreas Sandberg <andreas@sandberg.pp.se>

andysan commented:
As per previous discussion on #zephyr-lorawan, I have renamed the shell source to shell.c and switched to using an alias to discover the radio. I also updated the samples to use the same mechanism.

Mani-Sadhasivam commented on samples/drivers/lora/send/src/main.c:
```diff
@@ -10,6 +10,10 @@
 #include <zephyr.h>
 
+#define DEFAULT_RADIO_NODE DT_ALIAS(lora0)
+BUILD_ASSERT(DT_HAS_NODE(DEFAULT_RADIO_NODE), "No default LoRa radio");
```
"No default LoRa radio specified in DT"

Mani-Sadhasivam commented on samples/drivers/lora/receive/src/main.c:
```diff
@@ -10,6 +10,10 @@
 #include <zephyr.h>
 
+#define DEFAULT_RADIO_NODE DT_ALIAS(lora0)
+BUILD_ASSERT(DT_HAS_NODE(DEFAULT_RADIO_NODE), "No default LoRa radio");
```
"No default LoRa radio specified in DT"

Mani-Sadhasivam commented on boards/arm/96b_wistrio/96b_wistrio.dts:
```diff
@@ -61,7 +62,7 @@
 	cs-gpios = <&gpiob 0 0>;
 
-	sx1276@0 {
+	lora_radio: sx1276@0 {
```
Let's use `lora` as the node name. I've submitted a PR to add it in DT spec.

https://github.com/devicetree-org/devicetree-specification/pull/32

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+LOG_MODULE_REGISTER(lora_shell);
+
+#define DEFAULT_RADIO_NODE DT_ALIAS(lora0)
+BUILD_ASSERT(DT_HAS_NODE(DEFAULT_RADIO_NODE), "No default LoRa radio");
```
"No default LoRa radio specified in DT"

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+		} else if (!strcmp("sf", param)) {
+			if (value < SF_6 || value > SF_12) {
+				shell_error(shell,
+					    "Invalid spread factor: SF%i",
```
s/spread/spreading

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+		return ret;
+	}
+
+	shell_print(shell, "RSSI: %" PRIi16 " dBm, SNR:%" PRIi8 " dBm",
```
First print data and then RSSI,SNR values.

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+	return 0;
+}
+
+
```
Remove extra new line

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+BUILD_ASSERT(DT_HAS_NODE(DEFAULT_RADIO_NODE), "No default LoRa radio");
+#define DEFAULT_RADIO DT_LABEL(DEFAULT_RADIO_NODE)
+
+struct lora_modem_config modem_config = {
```
static?

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+#include <stdlib.h>
+
+#include <shell/shell.h>
+#include <drivers/lora.h>
```
Sort all includes alphabetically. No need to group them.

Mani-Sadhasivam commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,253 @@
+	static char buf[0xff];
+	struct device *dev;
+	int ret;
+	s32_t timeout = K_FOREVER;
```
We need to be careful here. In #24649, I've switched to new timeout API. And that got added to upcoming release milestone. So most probably it can go in before this PR. So please mention the dependency in PR description and rebase.

andysan replied:
I'm not sure why there is a dependency here. The LoRa API specifies that the timeout is in ms and that K_FOREVER is allowed. As far as I can tell, that's not changed by #24649.

Mani-Sadhasivam replied:
> However, after merging that pull request, we need to explicitly convert the timeout to a k_timeout_t.

This is the reason why I suggested to merge it after #24649.

Mani-Sadhasivam reviewed (changes requested):
Few misc changes required. Also this _should_ go after #24649

Mani-Sadhasivam commented:
Now this one can be merged without any issues. FYI @carlescufi

stephanosio commented:
> -:55: WARNING:SPDX_LICENSE_TAG: Missing or malformed SPDX-License-Identifier tag in line 1
> #55: FILE: drivers/lora/shell.c:1:
> +/*
> 
> -:58: WARNING:SPDX_LICENSE_TAG: Misplaced SPDX-License-Identifier tag - use line 1 instead
> #58: FILE: drivers/lora/shell.c:4:
> + * SPDX-License-Identifier: Apache-2.0

Please ignore these warnings (see #24921)

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+
+	return dev;
+}
+static int cmd_lora_conf(const struct shell *shell, size_t argc, char **argv)
```
Missing NL before this function.

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+static int cmd_lora_conf(const struct shell *shell, size_t argc, char **argv)
+{
+	if (argc < 2) {
+		shell_print(shell, DEFAULT_RADIO ":");
```
I would prefer to have:
```
if (argc < 2) {
	lora_config_dump(shell);
	return 0;
}
```
to reduce `cmd_lora_conf()` body a little bit.

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+	for (int i = 1; i < argc; i += 2) {
+		char *param = argv[i];
+		char *eptr;
+		int value;
```
What about introducing some helper function like `lora_conf_param_set(const struct shell *shell, const char *param, const char *value)` and then calling `lora_conf_param_set(shell, argv[i], argv[i + 1]);` in the loop? Currently there are too many lines of code and too many local variables in this function.

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+		return ret;
+	}
+
+	shell_print(shell, "Data: %s", buf);
```
We should use `shell_hexdump()` here for data printing. One thing that I noticed is that `shell_hexdump()` misses printing bytes as characters (in case buffer contains valid string), which is nicely done in `LOG_HEXDUMP_*()`.

andysan replied:
Hmm, this is a good point. It's unfortunately a choice between three bad options, shell_hexdump that doesn't play nicely with send (that sends a string), LOG_HEXDUMP that doesn't play nicely with the shell, and shell_hexdump that doesn't print binary. I think I prefer the current option for now, but it should be changed to shell_hexdump when/if it prints printable characters.

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+		return NULL;
+	}
+
+	if (cfg) {
```
I don't like the fact that `get_modem` applies config. I would rather call `device_get_binding()` separately and create a helper function for applying lora config (and checking inside `cfg->frequency == 0` a you do here).

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+
+	freq = atoi(argv[1]);
+	power = atoi(argv[2]);
+	duration = atoi(argv[3]);
```
In `cmd_lora_conf()` you use `strtol`. I think we should be consistent and use the same here.

mniestroj commented on drivers/lora/sx1276.c:
```diff
@@ -439,6 +439,13 @@ static int sx1276_lora_config(struct device *dev,
 	return 0;
 }
 
+int sx1276_lora_test_cw(struct device *dev, u32_t frequency,
```
`static`

mniestroj commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,247 @@
+		return -ENODEV;
+	}
+
+	ret = lora_recv(dev, buf, sizeof(buf), K_FOREVER, &rssi, &snr);
```
I don't like the K_FOREVER option. It will freeze console, with no option to get back from `lora_recv`. Explicit timeout would be better IMO.

andysan replied:
I have reintroduced the timeout option now that the API change has been merged.

jukkar commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,343 @@
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
+
+LOG_MODULE_REGISTER(lora_shell);
```
You can simplify this like this
```
LOG_MODULE_REGISTER(lora_shell, CONFIG_LORA_LOG_LEVEL);
```
Better use the lora debug level here.

jukkar commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,343 @@
+	[BW_500_KHZ] = 500,
+};
+
+
```
remove extra empty line

jukkar commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,343 @@
+		return lora_conf_dump(shell);
+	}
+
+	for (int i = 1; i < argc; i += 2) {
```
Please introduce variables at the beginning of the function

jukkar commented on drivers/lora/shell.c:
```diff
@@ -0,0 +1,343 @@
+			    size_t argc, char **argv)
+{
+	struct device *dev;
+	int ret = 0;
```
unnecessary assignment
